alf/nest/utils.py

```diff
@@ -176,6 +190,71 @@ def make_parallel(self, n):
         return NestMultiply(self._activation, "parallel_" + self._name)
 
 
+@alf.configurable
+class NestOuterProduct(NestCombiner):
+    def __init__(self,
+                 activation: Callable = None,
+                 batch_dims: int = 1,
+                 padding: bool = False,
+                 name: str = "NestOuterProduct"):
+        """Perform outer-product operations across a nested structure. Can be used
+        as a preprocessing combiner of a network.
+
+        Sometimes combining tensors using outer product might be more expressive
+        than concatenating, e.g., when one tensor is one-hot. See the discussions in
+
+        ::
+
+            "STOCHASTIC NEURAL NETWORKS FOR HIERARCHICAL REINFORCEMENT LEARNING",
+            Florensa, et al., ICLR 2017, https://arxiv.org/pdf/1704.03012.pdf.
+
+        In this implementation, we also support padding 1s to the tensors before
+        doing the outer product, essentially combining outer product and
+        concatenation together in one combiner.
+
+        .. warning::
+
+            Due to outer product, this combiner might result in a very long
+            output vector. Make sure to do the calculation before using it.
+
+        Args:
+            activation: optional activation function applied after the outer product.
+            batch_dims: number of batch dims. Default to 1. If the total input dim
+                is ``N``, then the last ``N-batch_dims`` will be flattened for
+                outer product.
+            padding: if True, each tensor will be padded by 1 before performing
+                outer product. When this flag is enabled, essentially it has
+                the effect of concatenation of all tensors in the output tensor.
+            name: name of the combiner
+        """
+        super(NestOuterProduct, self).__init__(name, batch_dims=batch_dims)
+        if activation is None:
+            activation = lambda x: x
+        self._activation = activation
+        self._padding = padding
+
+    def _combine_flat(self, tensors):
+        batch_shapes = tensors[0].shape[:self._batch_dims]
+        B = int(np.prod(batch_shapes))
+
+        tensors = [t.reshape(B, -1) for t in tensors]
```

Comment:
It would be better to add an assert to make sure all the tensors have the same "batch shape" to avoid a silent failure before this step, at the cost of `O(number of tensors)` (which I think is negligible).

Reply:
> It would be better to add an assert to make sure all the tensors have the same "batch shape" to avoid a silent failure before this step, at the cost of `O(number of tensors)` (which I think is negligible).

I think the batch shape can be different, as long as ``B`` is the same. If not, then reshape() will throw an error.